File: frontend/src/panels/map/globe.ts
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (C) 2026 Nurlan Urazkulov

/**
 * The globe: degrees onto the screen, and the hand turning it (D-319, wave 3).
 *
 * A surface node stands on its planet's sphere in degrees, and the map draws
 * the sphere as it is seen from above one point of it -- the **eye**: an
 * orthographic projection with north up, the eye's point at the origin of
 * the frame. Coordinates near the frame are small whatever the planet's
 * size, because the origin moves with the eye rather than staying at the
 * sphere's centre; that is what keeps a viewBox honest over ten orders of
 * magnitude (plan §2).
 *
 * What is behind the sphere is not drawn: `project` says so, and an edge is
 * a great-circle arc cut where it goes over the horizon. The hand turns the
 * globe by two angles and nothing else -- north stays up (D-237): "the mine
 * is north of the gate" stays true however one drags.
 *
 * Pure arithmetic, apart from React, so a test can turn the globe without a
 * pointer or a DOM.
 */

import type { Point } from "./model";

/** Where the eye stands over the sphere, in degrees. */
export type Eye = { lat: number; lon: number };
/** A point of the sphere, in degrees. */
export type Geo = { lat: number; lon: number };

/** Map units a metre of the surface makes: the flat map's own density at a
 *  city step, so a plot keeps the size it had before the globe. */
export const UNITS_PER_METRE = 5;
/** The eye stays this side of the pole so that "north up" keeps a meaning
 *  (plan §7, п. 2): at the pole every direction is south. */
export const LAST_LAT = 85;
/** How many pieces a great-circle arc is drawn in. */
export const ARC_STEPS = 16;

const RAD = Math.PI / 180;

/** The planet's radius in map units, from its radius in kilometres. */
export function radiusUnits(radiusKm: number): number {
  return radiusKm * 1000 * UNITS_PER_METRE;
}

/**
 * A point of the sphere as seen from the eye: where it lands on the frame,
 * and whether it faces the eye at all.
 */
export function project(
  eye: Eye,
  radius: number,
  point: Geo,
): { x: number; y: number; front: boolean } {
  const lat0 = eye.lat * RAD;
  const lat = point.lat * RAD;
  const dlon = (point.lon - eye.lon) * RAD;
  const cosc = Math.sin(lat0) * Math.sin(lat) + Math.cos(lat0) * Math.cos(lat) * Math.cos(dlon);
  return {
    x: radius * Math.cos(lat) * Math.sin(dlon),
    //: Screen y grows downwards; north is up.
    y: -radius * (Math.cos(lat0) * Math.sin(lat) - Math.sin(lat0) * Math.cos(lat) * Math.cos(dlon)),
    front: cosc >= 0,
  };
}

/** The point `share` of the way from `a` to `b` along the great circle. */
export function slerp(a: Geo, b: Geo, share: number): Geo {
  const toVec = (p: Geo) => {
    const lat = p.lat * RAD;
    const lon = p.lon * RAD;
    return [Math.cos(lat) * Math.cos(lon), Math.cos(lat) * Math.sin(lon), Math.sin(lat)];
  };
  const [ax, ay, az] = toVec(a);
  const [bx, by, bz] = toVec(b);
  const dot = Math.max(-1, Math.min(1, ax * bx + ay * by + az * bz));
  const omega = Math.acos(dot);
  if (omega < 1e-9) return { lat: a.lat, lon: a.lon };
  const wa = Math.sin((1 - share) * omega) / Math.sin(omega);
  const wb = Math.sin(share * omega) / Math.sin(omega);
  const x = wa * ax + wb * bx;
  const y = wa * ay + wb * by;
  const z = wa * az + wb * bz;
  return { lat: Math.atan2(z, Math.hypot(x, y)) / RAD, lon: Math.atan2(y, x) / RAD };
}

/**
 * The visible runs of the great-circle arc between two points: one polyline
 * where the whole arc faces the eye, none where none of it does, and pieces
 * where it dips over the horizon.
 */
export function arc(eye: Eye, radius: number, a: Geo, b: Geo, steps = ARC_STEPS): Point[][] {
  const runs: Point[][] = [];
  let run: Point[] = [];
  for (let i = 0; i <= steps; i++) {
    const here = project(eye, radius, slerp(a, b, i / steps));
    if (here.front) {
      run.push({ x: here.x, y: here.y });
    } else if (run.length) {
      runs.push(run);
      run = [];
    }
  }
  if (run.length) runs.push(run);
  return runs.filter((piece) => piece.length >= 2);
}

/**
 * The eye after the hand has dragged the ground by `dx`, `dy` map units:
 * the ground follows the hand, so the eye moves the other way -- west when
 * the ground goes east, south when it goes down. Two angles, north up.
 */
export function turn(eye: Eye, radius: number, dx: number, dy: number): Eye {
  const lat = Math.max(-LAST_LAT, Math.min(LAST_LAT, eye.lat + (dy / radius) / RAD));
  const stretch = Math.max(Math.cos(eye.lat * RAD), 1e-9);
  const lon = eye.lon - (dx / (radius * stretch)) / RAD;
  return { lat, lon: ((lon + 180) % 360 + 360) % 360 - 180 };
}

/** Every placed node of a scene as the eye sees it; what faces away is left out. */
export function projectAll(
  eye: Eye,
  radius: number,
  nodes: readonly { key: string; place?: Geo | { x: number; y: number } | null }[],
): Map<string, Point> {
  const out = new Map<string, Point>();
  for (const node of nodes) {
    if (!node.place || !("lat" in node.place)) continue;
    const seen = project(eye, radius, node.place);
    if (seen.front) out.set(node.key, { x: seen.x, y: seen.y });
  }
  return out;
}

File: backend/src/globe.py
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright (C) 2026 Nurlan Urazkulov

"""The sphere: where a point on a planet is, and how far two of them lie apart (D-319).

Outside `engine` on purpose, beside `astro` and `units`: this is arithmetic
with its own numbers -- the last latitude, the wrap of a longitude -- and the
rule modules hold no numbers of their own (D-065). The radius alone is the
vault's, and it is read here from the registry.

A node of a planet's surface stands at a latitude and a longitude, in degrees,
on a sphere whose radius comes from how much **land** the vault gives the world
(`planet.land_area_share` on `planet.earth_radius_km`, D-324) -- not from the
body it is in the sky, which is a different size on purpose. Everything
here is arithmetic over those two numbers: no session, no clock, no rounding
into map units -- the client fits degrees into whatever frame it draws.

Great-circle distances by the haversine formula; offsets on the tangent plane
are turned into degrees by the same radius. Good to a metre over the few
kilometres a city and its surroundings span, which is all the surface needs
until wave 5 of the plan reads the sky in the same kilometres.
"""

from __future__ import annotations

import math

import numpy as np

from src.constants import Constants
from src.constants import registry as R
from src.models.world import Planet
from src.units import METRES_PER_KM

#: A point on the sphere: latitude, longitude, degrees.
Geo = tuple[float, float]

#: The last usable latitude: "north up" is not defined on the pole itself, and
#: a seat pushed past it would come back round the other side.
LAST_LAT = 89.0
#: A full turn and a half turn of longitude, for wrapping.
HALF_TURN = 180.0
FULL_TURN = 360.0
QUARTER_TURN = 90.0
#: How many points along a straight way are read for what it crosses.
WAY_SAMPLES = 8
#: How many directions the compass is read in when looking round a point:
#: the ring a patch of ground is ranked against (landscape plan wave 7).
COMPASS_POINTS = 8
#: The golden angle: points fanned round a centre without a pattern.
GOLDEN_ANGLE = math.pi * (3 - math.sqrt(5))
#: Below this the cosine of the latitude is treated as at the pole.
COS_FLOOR = 1e-9


def radius_m(constants: Constants, planet: Planet) -> float:
    """The radius of the land one walks, in metres (D-324).

    Not the body in the sky: a world has two sizes now, and this is the one
    the map, the walk and the scout mean. It comes from how much **land** the
    vault gives the world -- a share of the Earth's surface, chosen for
    crowding and not for physics -- and a sphere of that area has this radius.
    A world the vault forgot gets the Earth's own land: absurdly large, and
    therefore noticed rather than quietly wrong.
    """
    share = float(constants[R.PLANET_LAND_AREA_SHARE].get(planet.value, 1.0))
    return math.sqrt(share) * float(constants[R.PLANET_EARTH_RADIUS_KM]) * METRES_PER_KM


def distance_m(radius: float, one: Geo, other: Geo) -> float:
    """The great-circle distance between two points of a sphere of that radius."""
    lat1, lon1 = math.radians(one[0]), math.radians(one[1])
    lat2, lon2 = math.radians(other[0]), math.radians(other[1])
    sin_lat = math.sin((lat2 - lat1) / 2)
    sin_lon = math.sin((lon2 - lon1) / 2)
    chord = sin_lat * sin_lat + math.cos(lat1) * math.cos(lat2) * sin_lon * sin_lon
    return 2 * radius * math.asin(min(1.0, math.sqrt(chord)))


def offset(radius: float, at: Geo, east_m: float, north_m: float) -> Geo:
    """The point `east_m` east and `north_m` north of `at`, on the tangent plane.

    The longitude step shrinks with the cosine of the latitude, so a step east
    is the same metres at the equator and near the pole; the latitude is held
    short of the pole itself.
    """
    lat = math.radians(at[0])
    d_lat = math.degrees(north_m / radius)
    d_lon = math.degrees(east_m / (radius * max(math.cos(lat), 1e-9)))
    new_lat = max(-LAST_LAT, min(LAST_LAT, at[0] + d_lat))
    new_lon = ((at[1] + d_lon + 180.0) % 360.0) - 180.0
    return (new_lat, new_lon)


def bearing(one: Geo, other: Geo) -> float:
    """The way from `one` to `other` as one sets out: degrees clockwise from north."""
    lat1, lat2 = math.radians(one[0]), math.radians(other[0])
    dlon = math.radians(wrap_lon(other[1] - one[1]))
    east = math.sin(dlon) * math.cos(lat2)
    north = math.cos(lat1) * math.sin(lat2) - math.sin(lat1) * math.cos(lat2) * math.cos(dlon)
    return math.degrees(math.atan2(east, north)) % FULL_TURN


def wrap_lon(lon: float) -> float:
    """A longitude brought into the half-open turn round the zero meridian."""
    return ((lon + HALF_TURN) % FULL_TURN) - HALF_TURN


def walk_between(a: Geo, b: Geo, share: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    """`between` for a whole walk at once: the points a share of the way from
    one to the other, the short way round."""
    turn = wrap_lon(b[1] - a[1])
    return a[0] + (b[0] - a[0]) * share, (a[1] + turn * share + HALF_TURN) % FULL_TURN - HALF_TURN


def between(a: Geo, b: Geo, share: float) -> Geo:
    """The point `share` of the way from `a` to `b`, the short way round."""
    dlon = wrap_lon(b[1] - a[1])
    return (a[0] + (b[0] - a[0]) * share, wrap_lon(a[1] + dlon * share))


def lon_stretch(lat: float) -> float:
    """How many degrees of longitude one degree of latitude's metres make here."""
    return 1 / max(math.cos(math.radians(lat)), COS_FLOOR)


def midpoint(a: float, b: float) -> float:
    """Halfway between two lengths."""
    return (a + b) / 2


def radius_of_area(area_m2: float) -> float:
    """The radius of the circle an area makes: a node's footprint on the map."""
    return math.sqrt(float(area_m2) / math.pi)
